File: sdks/rust/src/domain/decision.rs
//! Canonical evaluation decision (`spec/decision.schema.json`).

use super::errors::ErrorKind;
use super::types::{FlagMetadata, JsonValue};

/// Canonical reason strings (`spec/decision.schema.json`).
pub mod reason {
    pub const TARGETING_MATCH: &str = "TARGETING_MATCH";
    pub const SPLIT: &str = "SPLIT";
    pub const DISABLED: &str = "DISABLED";
    pub const DEFAULT: &str = "DEFAULT";
    pub const STALE: &str = "STALE";
    pub const CACHED: &str = "CACHED";
    pub const STATIC: &str = "STATIC";
    pub const ERROR: &str = "ERROR";
}

/// Result of a flag evaluation. Evaluation APIs return this, never raise
/// (`spec/control-points.md` "Return discipline — never throw into a read
/// path").
#[derive(Debug, Clone, PartialEq)]
pub struct Decision {
    pub value: JsonValue,
    pub variant: Option<String>,
    pub reason: String,
    pub error_code: Option<String>,
    pub error_message: Option<String>,
    pub error_kind: Option<ErrorKind>,
    pub flag_metadata: FlagMetadata,
}

impl Decision {
    /// A decision degraded to the caller's default with a no-frills reason
    /// (e.g. `DEFAULT`/`STATIC`) and no error.
    pub fn new(value: JsonValue, variant: Option<String>, reason: impl Into<String>) -> Self {
        Decision {
            value,
            variant,
            reason: reason.into(),
            error_code: None,
            error_message: None,
            error_kind: None,
            flag_metadata: FlagMetadata::new(),
        }
    }

    pub fn is_error(&self) -> bool {
        self.reason == reason::ERROR
    }
}

File: sdks/dart/lib/src/domain/decision.dart
import 'errors.dart';
import 'types.dart';

/// Canonical reason strings (`spec/decision.schema.json`).
///
/// `default` and `static` are reserved words in Dart, hence
/// [defaultReason] / [staticReason] — the wire names are unchanged.
enum DecisionReason {
  targetingMatch('TARGETING_MATCH'),
  split('SPLIT'),
  disabled('DISABLED'),
  defaultReason('DEFAULT'),
  stale('STALE'),
  cached('CACHED'),
  staticReason('STATIC'),
  error('ERROR');

  const DecisionReason(this.wireName);

  /// The reason string `contracts/` fixtures and the wire protocol use.
  final String wireName;

  /// Parses a wire name; `null` when it is not one of the eight.
  static DecisionReason? fromWireName(String raw) {
    for (final reason in values) {
      if (reason.wireName == raw) {
        return reason;
      }
    }
    return null;
  }
}

/// Result of a flag evaluation. Evaluation APIs return this, never throw
/// (`spec/control-points.md` "Return discipline — never throw into a read
/// path").
class Decision {
  const Decision({
    required this.value,
    this.variant,
    required this.reason,
    this.errorCode,
    this.errorMessage,
    this.errorKind,
    this.flagMetadata = const <String, Object?>{},
  });

  final JsonValue value;
  final String? variant;
  final DecisionReason reason;
  final String? errorCode;
  final String? errorMessage;
  final ErrorKind? errorKind;
  final FlagMetadata flagMetadata;

  bool get isError => reason == DecisionReason.error;

  /// Plain-JSON snapshot in the shape `contracts/` fixtures compare against.
  Map<String, Object?> toJson() => <String, Object?>{
    'value': value,
    'variant': variant,
    'reason': reason.wireName,
    'errorCode': errorCode,
    'errorMessage': errorMessage,
    'flagMetadata': flagMetadata,
  };

  @override
  String toString() =>
      'Decision(${reason.wireName} value=$value variant=$variant '
      'errorKind=${errorKind?.wireName} metadata=$flagMetadata)';
}
